File: test/scripts/05-branch-misprediction/correct.py
#!/usr/bin/env python3
"""
CORRECT: Sort data first - predictable branches. Or use branchless.
Resume: "Branch Prediction"
"""
import os
import signal
import time
N = 5_000_000
DURATION_SEC = int(os.environ.get("DURATION_SEC", "300"))
stop = False

# Sorted: evens first, then odds -> predictable
def sum_predictable(data):
    total = 0
    for x in data:
        if x % 2 == 0:
            total += x
        else:
            total -= x
    return total

def main():
    global stop
    data = list(range(100)) * (N // 100)
    data.sort(key=lambda x: x % 2)  # Evens first -> predictable
    print("CORRECT: Sorted data -> predictable branches")
    signal.signal(signal.SIGINT, lambda s, f: globals().__setitem__('stop', True))
    signal.signal(signal.SIGTERM, lambda s, f: globals().__setitem__('stop', True))
    start = time.perf_counter()
    laps = 0
    while not stop and (time.perf_counter() - start) < DURATION_SEC:
        sum_predictable(data)
        laps += 1
    print(f"Stopped: {laps} laps")

if __name__ == "__main__":
    main()

File: test/scripts/05-branch-misprediction/correct.js
const N = 5_000_000;
const DURATION_SEC = parseInt(process.env.DURATION_SEC || '300', 10);
const data = Array.from({ length: 100 }, (_, i) => i);
const arr = [];
for (let i = 0; i < N; i++) arr.push(data[i % 100]);
arr.sort((a, b) => (a % 2) - (b % 2));  // Evens first -> predictable
function sumPredictable(a) {
  let total = 0;
  for (let i = 0; i < a.length; i++) {
    if (a[i] % 2 === 0) total += a[i];
    else total -= a[i];
  }
  return total;
}
let laps = 0;
const start = Date.now();
console.log('CORRECT: Sorted -> predictable');
const id = setInterval(() => {
  if (Date.now() - start >= DURATION_SEC * 1000) {
    clearInterval(id);
    console.log('Stopped:', laps, 'laps');
    return;
  }
  sumPredictable(arr);
  laps++;
}, 0);
process.on('SIGINT', () => { clearInterval(id); process.exit(0); });
